File: kreuzberg/packages/r/R/extraction.R
#' Extract content from a file (synchronous)
#'
#' @param path Path to the file.
#' @param mime_type Optional MIME type override.
#' @param config Optional extraction configuration from \code{extraction_config()}.
#' @return A \code{kreuzberg_result} object.
#' @export
extract_file_sync <- function(path, mime_type = NULL, config = NULL) {
  stopifnot(is.character(path), length(path) == 1L)
  if (!file.exists(path)) stop("File not found: ", path, call. = FALSE)
  if (!is.null(mime_type)) stopifnot(is.character(mime_type), length(mime_type) == 1L)
  config_json <- if (!is.null(config)) jsonlite::toJSON(config, auto_unbox = TRUE) else NULL
  result <- check_native_result(extract_file_sync_native(path, mime_type, config_json))
  as_kreuzberg_result(result)
}

#' Extract content from a file (async, blocks in R)
#'
#' @param path Path to the file.
#' @param mime_type Optional MIME type override.
#' @param config Optional extraction configuration from \code{extraction_config()}.
#' @return A \code{kreuzberg_result} object.
#' @export
extract_file <- function(path, mime_type = NULL, config = NULL) {
  stopifnot(is.character(path), length(path) == 1L)
  if (!file.exists(path)) stop("File not found: ", path, call. = FALSE)
  if (!is.null(mime_type)) stopifnot(is.character(mime_type), length(mime_type) == 1L)
  config_json <- if (!is.null(config)) jsonlite::toJSON(config, auto_unbox = TRUE) else NULL
  result <- check_native_result(extract_file_native(path, mime_type, config_json))
  as_kreuzberg_result(result)
}

#' Extract content from raw bytes (synchronous)
#'
#' @param data Raw vector of bytes.
#' @param mime_type MIME type of the data.
#' @param config Optional extraction configuration from \code{extraction_config()}.
#' @return A \code{kreuzberg_result} object.
#' @export
extract_bytes_sync <- function(data, mime_type, config = NULL) {
  stopifnot(is.raw(data))
  stopifnot(is.character(mime_type), length(mime_type) == 1L)
  config_json <- if (!is.null(config)) jsonlite::toJSON(config, auto_unbox = TRUE) else NULL
  result <- check_native_result(extract_bytes_sync_native(data, mime_type, config_json))
  as_kreuzberg_result(result)
}

#' Extract content from raw bytes (async, blocks in R)
#'
#' @param data Raw vector of bytes.
#' @param mime_type MIME type of the data.
#' @param config Optional extraction configuration from \code{extraction_config()}.
#' @return A \code{kreuzberg_result} object.
#' @export
extract_bytes <- function(data, mime_type, config = NULL) {
  stopifnot(is.raw(data))
  stopifnot(is.character(mime_type), length(mime_type) == 1L)
  config_json <- if (!is.null(config)) jsonlite::toJSON(config, auto_unbox = TRUE) else NULL
  result <- check_native_result(extract_bytes_native(data, mime_type, config_json))
  as_kreuzberg_result(result)
}

#' Render a single PDF page as a PNG image
#'
#' @param path Path to the PDF file.
#' @param page_index Zero-based page index.
#' @param dpi Rendering resolution (default 150).
#' @return A raw vector containing PNG-encoded bytes.
#' @export
render_pdf_page <- function(path, page_index, dpi = 150L) {
  stopifnot(is.character(path), length(path) == 1L)
  if (!file.exists(path)) stop("File not found: ", path, call. = FALSE)
  stopifnot(is.numeric(page_index), length(page_index) == 1L)
  if (page_index < 0L) stop("page_index must be non-negative", call. = FALSE)
  stopifnot(is.numeric(dpi), length(dpi) == 1L)
  check_native_result(render_pdf_page_native(path, as.integer(page_index), as.integer(dpi)))
}

#' Iterate over PDF pages, calling a callback per page
#'
#' Renders each page one at a time and invokes the callback with (page_index, png_raw).
#' This keeps memory usage low for large PDFs because only one page is in memory at a time.
#'
#' @param path Path to the PDF file.
#' @param dpi Rendering resolution (default 150).
#' @param callback A function taking two arguments: page_index (integer) and png_raw (raw vector).
#' @return Invisible NULL. Called for side effects.
#' @export
render_pdf_pages_iter <- function(path, dpi = 150L, callback) {
  stopifnot(is.character(path), length(path) == 1L)
  if (!file.exists(path)) stop("File not found: ", path, call. = FALSE)
  stopifnot(is.numeric(dpi), length(dpi) == 1L)
  stopifnot(is.function(callback))

  page_index <- 0L
  repeat {
    result <- tryCatch(
      render_pdf_page(path, page_index, dpi),
      error = function(e) {
        if (grepl("not found|out of bounds|Page.*not found", conditionMessage(e), ignore.case = TRUE)) {
          return(NULL) # end of pages
        }
        stop(e) # re-raise real errors
      }
    )
    if (is.null(result)) break
    callback(page_index, result)
    page_index <- page_index + 1L
  }
  invisible(NULL)
}

File: kreuzberg/packages/r/src/rust/src/extraction.rs
//! File and bytes extraction functions (sync + async via Tokio)

use crate::config::parse_config;
use crate::error::{kreuzberg_error, to_r_error};
use crate::result::extraction_result_to_list;
use extendr_api::prelude::*;

pub fn extract_file_sync_impl(path: &str, mime_type: Nullable<&str>, config_json: Nullable<&str>) -> extendr_api::Result<List> {
    #[cfg(not(target_arch = "wasm32"))]
    {
        let config = parse_config(config_json)?;
        let mime = match mime_type {
            Nullable::NotNull(m) => Some(m),
            Nullable::Null => None,
        };
        let result = kreuzberg::extract_file_sync(path, mime, &config).map_err(kreuzberg_error)?;
        extraction_result_to_list(result)
    }
    #[cfg(target_arch = "wasm32")]
    {
        let _ = (path, mime_type, config_json);
        Err("File extraction is not supported on WebAssembly".into())
    }
}

pub fn extract_file_impl(path: &str, mime_type: Nullable<&str>, config_json: Nullable<&str>) -> extendr_api::Result<List> {
    #[cfg(not(target_arch = "wasm32"))]
    {
        let config = parse_config(config_json)?;
        let mime = match mime_type {
            Nullable::NotNull(m) => Some(m),
            Nullable::Null => None,
        };
        let runtime = tokio::runtime::Runtime::new().map_err(to_r_error)?;
        let result = runtime
            .block_on(async { kreuzberg::extract_file(path, mime, &config).await })
            .map_err(kreuzberg_error)?;
        extraction_result_to_list(result)
    }
    #[cfg(target_arch = "wasm32")]
    {
        let _ = (path, mime_type, config_json);
        Err("Async file extraction is not supported on WebAssembly".into())
    }
}

pub fn extract_bytes_sync_impl(data: Raw, mime_type: &str, config_json: Nullable<&str>) -> extendr_api::Result<List> {
    let config = parse_config(config_json)?;
    let bytes = data.as_slice();
    let result = kreuzberg::extract_bytes_sync(bytes, mime_type, &config).map_err(kreuzberg_error)?;
    extraction_result_to_list(result)
}

pub fn render_pdf_page_impl(path: &str, page_index: i32, dpi: i32) -> extendr_api::Result<Raw> {
    #[cfg(not(target_arch = "wasm32"))]
    {
        if page_index < 0 {
            return Err("page_index must be non-negative".into());
        }
        let pdf_bytes = std::fs::read(path).map_err(to_r_error)?;
        let dpi_opt = if dpi <= 0 { None } else { Some(dpi) };
        let png = kreuzberg::pdf::render_pdf_page_to_png(&pdf_bytes, page_index as usize, dpi_opt, None)
            .map_err(to_r_error)?;
        Ok(Raw::from_bytes(&png))
    }
    #[cfg(target_arch = "wasm32")]
    {
        let _ = (path, page_index, dpi);
        Err("PDF rendering is not supported on WebAssembly".into())
    }
}

pub fn extract_bytes_impl(data: Raw, mime_type: &str, config_json: Nullable<&str>) -> extendr_api::Result<List> {
    #[cfg(not(target_arch = "wasm32"))]
    {
        let config = parse_config(config_json)?;
        let bytes = data.as_slice();
        let runtime = tokio::runtime::Runtime::new().map_err(to_r_error)?;
        let result = runtime
            .block_on(async { kreuzberg::extract_bytes(bytes, mime_type, &config).await })
            .map_err(kreuzberg_error)?;
        extraction_result_to_list(result)
    }
    #[cfg(target_arch = "wasm32")]
    {
        extract_bytes_sync_impl(data, mime_type, config_json)
    }
}
